File: src/tools/edit-case.ts
import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { FogBugzClient } from "../fogbugz-client.js";

export function registerEditCase(
  server: McpServer,
  client: FogBugzClient,
) {
  server.tool(
    "edit_case",
    "Edit fields on a FogBugz case: title, project, area, milestone, priority, category, tags, kanban column, parent case, due date, and/or add a comment.",
    {
      case_number: z.number().describe("The FogBugz case number to edit"),
      title: z.string().optional().describe("New title for the case"),
      project: z.string().optional().describe("Project name to move the case to"),
      area: z.string().optional().describe("Area name"),
      milestone: z.string().optional().describe("Milestone name"),
      priority: z.string().optional().describe("Priority name (e.g. 'Must Fix', 'Fix If Time')"),
      category: z.string().optional().describe("Category name (e.g. 'Bug', 'Feature')"),
      tags: z
        .array(z.string())
        .optional()
        .describe("Tags to set on the case (replaces all existing tags)"),
      kanban_column: z.string().optional().describe("Kanban column name"),
      parent_case: z.number().optional().describe("Parent case number"),
      due_date: z.string().optional().describe("Due date in ISO 8601 format (e.g. '2025-12-31T00:00:00Z')"),
      comment: z.string().optional().describe("Comment to add to the case"),
    },
    { destructiveHint: true },
    async (params) => {
      const fields: Record<string, unknown> = {};

      if (params.title !== undefined) fields.sTitle = params.title;
      if (params.project !== undefined) fields.sProject = params.project;
      if (params.area !== undefined) fields.sArea = params.area;
      if (params.milestone !== undefined) fields.sFixFor = params.milestone;
      if (params.priority !== undefined) fields.sPriority = params.priority;
      if (params.category !== undefined) fields.sCategory = params.category;
      if (params.tags !== undefined) fields.sTags = params.tags.join(",");
      if (params.kanban_column !== undefined)
        fields.sKanbanColumn = params.kanban_column;
      if (params.parent_case !== undefined)
        fields.ixBugParent = params.parent_case;
      if (params.due_date !== undefined) fields.dtDue = params.due_date;
      if (params.comment !== undefined) fields.sEvent = params.comment;

      const result = await client.edit(params.case_number, fields);
      const caseUrl = client.getCaseUrl(result.case.ixBug);

      const changedFields = Object.keys(fields)
        .filter((k) => k !== "sEvent")
        .join(", ");
      const summary = changedFields
        ? `Updated fields: ${changedFields}`
        : "Added comment";

      return {
        content: [
          {
            type: "text",
            text: `Case ${result.case.ixBug} edited successfully.\n${summary}\nLink: ${caseUrl}`,
          },
        ],
      };
    },
  );
}

File: python/src/fogbugz_mcp/tools/edit_case.py
from __future__ import annotations

from typing import TYPE_CHECKING

from mcp.server.fastmcp import FastMCP

if TYPE_CHECKING:
    from ..fogbugz_client import FogBugzClient


def register_edit_case(mcp: FastMCP, client: FogBugzClient) -> None:
    @mcp.tool(
        name="edit_case",
        description=(
            "Edit fields on a FogBugz case: title, project, area, milestone, priority, "
            "category, tags, kanban column, parent case, due date, and/or add a comment."
        ),
        annotations={"destructiveHint": True},
    )
    async def edit_case(
        case_number: int,
        title: str | None = None,
        project: str | None = None,
        area: str | None = None,
        milestone: str | None = None,
        priority: str | None = None,
        category: str | None = None,
        tags: list[str] | None = None,
        kanban_column: str | None = None,
        parent_case: int | None = None,
        due_date: str | None = None,
        comment: str | None = None,
    ) -> str:
        """Edit fields on a FogBugz case."""
        fields: dict = {}
        if title is not None:
            fields["sTitle"] = title
        if project is not None:
            fields["sProject"] = project
        if area is not None:
            fields["sArea"] = area
        if milestone is not None:
            fields["sFixFor"] = milestone
        if priority is not None:
            fields["sPriority"] = priority
        if category is not None:
            fields["sCategory"] = category
        if tags is not None:
            fields["sTags"] = ",".join(tags)
        if kanban_column is not None:
            fields["sKanbanColumn"] = kanban_column
        if parent_case is not None:
            fields["ixBugParent"] = parent_case
        if due_date is not None:
            fields["dtDue"] = due_date
        if comment is not None:
            fields["sEvent"] = comment

        result = await client.edit(case_number, fields)
        case_url = client.get_case_url(result["case"]["ixBug"])

        changed = [k for k in fields if k != "sEvent"]
        summary = f"Updated fields: {', '.join(changed)}" if changed else "Added comment"

        return f"Case {result['case']['ixBug']} edited successfully.\n{summary}\nLink: {case_url}"
